pkg/sql/regionliveness/prober.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+var RegionLivenessEnabled = settings.RegisterBoolSetting(settings.ApplicationLevel,
+	"sql.region_liveness.enabled",
+	"enables region liveness for system databases (experimental)",
+	false, /* disabled */
+	settings.WithVisibility(settings.Reserved))
+
+// LiveRegions are regions which are currently still avaialble,
+// and not quarantined due to expiration.
+type LiveRegions map[string]struct{}
+
+// Prober used to determine the set of regions which are still alive.
+type Prober interface {
+	// ProbeLiveness can be used after a timeout to label a regions as unavailable.
+	ProbeLiveness(ctx context.Context, region string) error
+	// QueryLiveness can be used to get the list of regions which are currently
+	// accessible.
+	QueryLiveness(ctx context.Context, txn *kv.Txn) (LiveRegions, error)
+}
+
+// RegionProvider abstracts the lookup of regions (see regions.Provider).
+type RegionProvider interface {
+	// GetRegions provides access to the set of regions available to the
+	// current tenant.
+	GetRegions(ctx context.Context) (*serverpb.RegionsResponse, error)
+}
+
+type RegionProviderFactory func(txn *kv.Txn) RegionProvider
+
+type livenessProber struct {
+	db                    isql.DB
+	regionProviderFactory RegionProviderFactory
+	settings              *clustersettings.Settings
+}
+
+// NewLivenessProber creates a new region liveness prober.
+func NewLivenessProber(
+	db isql.DB, regionProviderFactory RegionProviderFactory, settings *clustersettings.Settings,
+) Prober {
+	return &livenessProber{
+		db:                    db,
+		regionProviderFactory: regionProviderFactory,
+		settings:              settings,
+	}
+}
+
+// ProbeLiveness implements Prober.
+func (l *livenessProber) ProbeLiveness(ctx context.Context, region string) error {
+	// If region liveness is disabled then nothing to do.
+	if !RegionLivenessEnabled.Get(&l.settings.SV) {
+		return nil
+	}
+	const probeQuery = `
+SELECT * FROM system.sql_instances WHERE crdb_region=$1::system.crdb_internal_region
+`
+	err := timeutil.RunWithTimeout(ctx, "probe-liveness", time.Second*15,
+		func(ctx context.Context) error {
+			return l.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+				_, err := txn.Exec(ctx, "probe-sql-instance", txn.KV(), probeQuery, region)
+				return err
+			})
+		})
+
+	// Region is alive or we hit some other error.
+	if err == nil ||
+		!IsQueryTimeoutErr(err) {
+		return err
+	}
+
+	// Region has gone down, set the unavailable_at time on it
+	return l.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		defaultTTL := slinstance.DefaultTTL.Get(&l.settings.SV)
```

Comment:
Something to consider for the future: we could create a sql function that wraps this (e.g `crdb_internal.set_region_unavailable(region)`). That allows us to write a roachprod test that randomly marks regions as unavailable and checks to make sure the sql_liveness expires_at timestamp is always less than the unavailable_at timestamp.

Reply:
This is a good idea and I think I'll pick this up in the next PR for lease table